### PR TITLE
Make it possible to override fromGlobals easily

### PR DESCRIPTION
**Pull request by sasanrose**
If I wanted to use a custom ServerRequest which extends Zend diactoros ServerRequest and still use the ServerRequestFactory::fromGlobals it was not possible without duplicating some methods and code. This change make it easier to extend ServerRequestFactory, override fromGlobals method and as a result use your own custom version of ServerRequest.

### CODE REVIEW

**Comment by sasanrose** [comment]
A test is failing in master branch which fails in my branch as well. I did not know how to fix it properly. The test is not related to this pull request.

File: src/ServerRequestFactory.php [hunk 1]
```diff
@@ -472,7 +510,7 @@ private static function normalizeNestedFileSpec(array $files = [])
      * @param array $server
      * @return string
      */
-    private static function marshalProtocolVersion(array $server)
+    public static function marshalProtocolVersion(array $server)
```
Comment (Ocramius) [inline src/ServerRequestFactory.php]: To be reverted - this is `private` API for which we shouldn't provide BC guarantees

File: src/ServerRequestFactory.php [hunk 2]
```diff
@@ -497,7 +535,7 @@ private static function marshalProtocolVersion(array $server)
      * @param $cookieHeader
      * @return array
      */
-    private static function parseCookieHeader($cookieHeader)
+    public static function parseCookieHeader($cookieHeader)
```
Comment (Ocramius) [inline src/ServerRequestFactory.php]: To be reverted - this is `private` API for which we shouldn't provide BC guarantees

**Comment by Ocramius** [comment]
Not closing yet: need an opinion from @weierophinney.

I don't think C&P is as bad as it sounds, compared to the introduced BC boundaries.

**Comment by weierophinney** [comment]
I tend to agree with @Ocramius here; the factory provided is specific to our `ServerRequest` implementation, and if you want to change how it does its work, you should create your own implementation, and/or provide a pull request with the changes and an argument for why we should alter how we currently do the work.

The various aspects of this particular factory are the result of our experiences in managing incoming requests since the early days of ZF1 through the present, and have been vetted a ton. Making it possible to extend the factory in order to alter that behavior will lead to more maintenance issues for us down the line. Forcing users to create their own factories in those situations will force users to fully consider the potential impact of the changes they want to introduce.

**Comment by sasanrose** [comment]
@weierophinney @Ocramius Sorry for late reply. I see that you guys closed this request and I understand your concerns. But I believe one of the strengths of zend-diactoros is the ServerRequestFactory class and all it's useful methods. A lot of them can be changed to be reused without even having any effects on the performance or behaviour.
Imagine I have a ServerRequest that goes through a bunch of middlewares and then passed to my controllers. In one of my middlewares I want to parse some query parameters and store it in my ServerRequest (i.e withKeys or withAction) and then receive them in my controllers (getKeys or getActions). Just for this I have to go through the re-implementing or duplicating a lot of ServerRequestFactory methods. While I could reuse it easily.

**Comment by Ocramius** [comment]
> A lot of them can be changed to be reused without even having any effects on the performance or behaviour.

The problems are:

 * stability
 * BC boundaries
 * coupling

Inheritance is the worst kind of coupling after shared singleton state.

 > Imagine I have a ServerRequest that goes through a bunch of middlewares and then passed to my controllers. In one of my middlewares I want to parse some query parameters and store it in my ServerRequest (i.e withKeys or withAction) and then receive them in my controllers (getKeys or getActions). Just for this I have to go through the re-implementing or duplicating a lot of ServerRequestFactory methods. While I could reuse it easily.

You are probably looking for `withAttribute`. The correct approach to this is to use that, even if it has less type safety. You can see examples of how that is correctly used in https://github.com/psr7-sessions/storageless/blob/e9d5b23503e68b10d19ac637731fef86ebb9f4ca/src/Storageless/Http/SessionMiddleware.php#L184 or https://github.com/dflydev/dflydev-fig-cookies.

**Comment by weierophinney** [comment]
@sasanrose For a reference, here's an excellent blog post detailing some of the issues we have with modifying classes to open the API: https://medium.com/@nicolopigna/inheritance-is-evil-stop-using-it-6c4f1caf5117

Essentially, what happens if we change the behavior of one of the methods you propose changing from `private` to `protected`? What if you depended on the previous behavior? We've now broken _your_ code by making a change in _our_ code.

Yes, this means more work for _you_ if you want to extend the class and use the existing behavior. However, it also isolates your implementation from upstream changes, ensuring it works consistently for your own needs. If we report bugfixes that you are interested in, you can apply the same patches to your code; however, if those bugfixes would represent a BC break for you, you're free to ignore them.

**Comment by sasanrose** [comment]
@weierophinney Fair enough. Thanks for the explanation and the reference

**Pull request closed without merging**